File: backend/msreserva/src/main/java/com/dac/msreserva/RabbitMQConfig/RabbitMQConfig.java
package com.dac.msreserva.RabbitMQConfig;

import org.springframework.amqp.core.*;
import org.springframework.amqp.rabbit.config.SimpleRabbitListenerContainerFactory;
import org.springframework.amqp.rabbit.connection.ConnectionFactory;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

@Configuration
public class RabbitMQConfig {

    private static final String DEFAULT_ROUTING_KEY = "reserva";

    @Bean
    public Queue novasReservas() {
        return new Queue("criareserva.reserva");
    }

    @Bean
    public DirectExchange sagaCriarReserva() {
        return new DirectExchange("criareserva");
    }

    @Bean
    public Queue reservasCQRSgravacao() {
        return new Queue("cqrs.gravacao");
    }

    @Bean
    public Queue reservasCQRSedicao() {
        return new Queue("cqrs.edicao");
    }

    @Bean
    public DirectExchange reservasCQRSExchange() {
        return new DirectExchange("cqrs");
    }

    @Bean
    public Queue cancelarReserva() {
        return new Queue("cancelareserva.reserva");
    }

    @Bean
    public DirectExchange sagaCancelarReserva() {
        return new DirectExchange("cancelareserva");
    }

    @Bean
    public Queue cancelarVoo() {
        return new Queue("cancelavoo.reserva");
    }

    @Bean
    public DirectExchange sagaCancelarVoo() {
        return new DirectExchange("cancelavoo");
    }

    @Bean
    public Queue realizarVoo() {
        return new Queue("realizavoo.reserva");
    }

    @Bean
    public DirectExchange sagaRealizarVoo() {
        return new DirectExchange("realizavoo");
    }

    @Bean
    public Binding bindingCriarReserva(DirectExchange sagaCriarReserva, Queue novasReservas) {
        return BindingBuilder.bind(novasReservas)
                .to(sagaCriarReserva)
                .with(DEFAULT_ROUTING_KEY);
    }

    @Bean
    public Binding bindingCQRSReservaGravacao(DirectExchange reservasCQRSExchange, Queue reservasCQRSgravacao) {
        return BindingBuilder.bind(reservasCQRSgravacao)
                .to(reservasCQRSExchange)
                .with("gravacao");
    }

    @Bean
    public Binding bindingCQRSReservaEdicao(DirectExchange reservasCQRSExchange, Queue reservasCQRSedicao) {
        return BindingBuilder.bind(reservasCQRSedicao)
                .to(reservasCQRSExchange)
                .with("edicao");
    }

    @Bean
    public Binding bindingCancelarReserva(DirectExchange sagaCancelarReserva, Queue cancelarReserva) {
        return BindingBuilder.bind(cancelarReserva)
                .to(sagaCancelarReserva)
                .with(DEFAULT_ROUTING_KEY);
    }

    @Bean
    public Binding bindingCancelarVoo(DirectExchange sagaCancelarVoo, Queue cancelarVoo) {
        return BindingBuilder.bind(cancelarVoo)
                .to(sagaCancelarVoo)
                .with(DEFAULT_ROUTING_KEY);
    }

    @Bean
    public Binding bindingRealizarVoo(DirectExchange sagaRealizarVoo, Queue realizarVoo) {
        return BindingBuilder.bind(realizarVoo)
                .to(sagaRealizarVoo)
                .with(DEFAULT_ROUTING_KEY);
    }

    @Bean
    public SimpleRabbitListenerContainerFactory rabbitListenerContainerFactory(ConnectionFactory connectionFactory) {
        SimpleRabbitListenerContainerFactory factory = new SimpleRabbitListenerContainerFactory();
        factory.setConnectionFactory(connectionFactory);
        factory.setDefaultRequeueRejected(false);
        return factory;
    }
}

File: backend/reserva-service/src/main/kotlin/br/ufpr/dac/reserva_service/RabbitMQConfig.kt
package br.ufpr.dac.reserva_service

import org.springframework.amqp.core.Binding
import org.springframework.amqp.core.BindingBuilder
import org.springframework.amqp.core.DirectExchange
import org.springframework.amqp.core.Queue
import org.springframework.amqp.rabbit.config.SimpleRabbitListenerContainerFactory
import org.springframework.amqp.rabbit.connection.ConnectionFactory
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration

@Configuration
class RabbitMQConfig {
    private final val DEFAULT_ROUTING_KEY = "reserva"

    @Bean
    fun novasReservas(): Queue = Queue("emiratads.criareserva.reserva")

    @Bean
    fun sagaCriarReserva(): DirectExchange = DirectExchange("emiratads.criareserva")

    @Bean
    fun reservasCQRSgravacao(): Queue = Queue("emiratads.cqrs.gravacao")

    @Bean
    fun reservasCQRSedicao(): Queue = Queue("emiratads.cqrs.edicao")

    @Bean
    fun reservasCQRSExchange(): DirectExchange = DirectExchange("emiratads.cqrs")

    @Bean
    fun cancelarReserva(): Queue = Queue("emiratads.cancelareserva.reserva")

    @Bean
    fun sagaCancelarReserva(): DirectExchange = DirectExchange("emiratads.cancelareserva")

    @Bean
    fun cancelarVoo(): Queue = Queue("emiratads.cancelavoo.reserva")

    @Bean
    fun sagaCancelarVoo(): DirectExchange = DirectExchange("emiratads.cancelavoo")

    @Bean
    fun realizarVoo(): Queue = Queue("emiratads.realizavoo.reserva")

    @Bean
    fun sagaRealizarVoo(): DirectExchange = DirectExchange("emiratads.realizavoo")

    @Bean
    fun bindingCriarReserva(
        sagaCriarReserva: DirectExchange,
        novasReservas: Queue
    ): Binding {
        return BindingBuilder.bind(novasReservas)
            .to(sagaCriarReserva)
            .with(DEFAULT_ROUTING_KEY)
    }

    @Bean
    fun bindingCQRSReservaGravacao(
        reservasCQRSExchange: DirectExchange,
        reservasCQRSgravacao: Queue
    ): Binding {
        return BindingBuilder.bind(reservasCQRSgravacao)
            .to(reservasCQRSExchange)
            .with("gravacao")
    }

    @Bean
    fun bindingCQRSReservaEdicao(
        reservasCQRSExchange: DirectExchange,
        reservasCQRSedicao: Queue
    ): Binding {
        return BindingBuilder.bind(reservasCQRSedicao)
            .to(reservasCQRSExchange)
            .with("edicao")
    }

    @Bean
    fun bindingCancelarReserva(
        sagaCancelarReserva: DirectExchange,
        cancelarReserva: Queue
    ): Binding {
        return BindingBuilder.bind(cancelarReserva)
            .to(sagaCancelarReserva)
            .with(DEFAULT_ROUTING_KEY)
    }

    @Bean
    fun bindingCancelarVoo(
        sagaCancelarVoo: DirectExchange,
        cancelarVoo: Queue
    ): Binding {
        return BindingBuilder.bind(cancelarVoo)
            .to(sagaCancelarVoo)
            .with(DEFAULT_ROUTING_KEY)
    }

    @Bean
    fun bindingRealizarVoo(
        sagaRealizarVoo: DirectExchange,
        realizarVoo: Queue
    ): Binding {
        return BindingBuilder.bind(realizarVoo)
            .to(sagaRealizarVoo)
            .with(DEFAULT_ROUTING_KEY)
    }

    @Bean
    fun rabbitListenerContainerFactory(connectionFactory: ConnectionFactory): SimpleRabbitListenerContainerFactory {
        val factory = SimpleRabbitListenerContainerFactory();
        factory.setConnectionFactory(connectionFactory)
        factory.setDefaultRequeueRejected(false)
        return factory
    }
}